clients/src/main/java/org/apache/kafka/clients/NetworkClient.java

```diff
@@ -639,19 +647,31 @@ private void maybeUpdate(long now, Node node) {
                 ClientRequest clientRequest = request(now, nodeConnectionId, metadataRequest);
                 log.debug("Sending metadata request {} to node {}", metadataRequest, node.id());
                 doSend(clientRequest, now);
-            } else if (connectionStates.canConnect(nodeConnectionId, now)) {
+                return requestTimeoutMs;
+            }
+
+            if (lastConnectAttemptNodeId != null &&
```

Comment:
Let me clarify my expected behavior of `metadataUpdate()`:

1. No matter if there's an ongoing connection establishment, if there's a node which can be used to send a metadata update request, we want to use it.
2. If there's no node which is available to send metadata request immediately, we have to prepare a new connection which will be usable to send metadata request in few rounds later.
   However, if there's an ongoing connection establishment which is started by the previous round, we want to wait it completes in order to avoid establishing unnecessary connections additionally.
   The exceptional case is that the ongoing connection establishment has already failed. In this case it's meangless to wait for the last attempt completes so we should move on next try immediately. I guess lastConnectAttemptNodeId is necessary to check whether the last attempt failed or not.
3. If the last connect attempt failed or `reconnect.backoff.ms` elapsed since then, we can try to connect next candidate.

I'm guessing maybe we have different thoughts on the point: "Is it okay to wait reconnect.backoff.ms for single connect attempt even though it failed immediately?".
I'm thinking we shouldn't apply `reconnect.backoff.ms` for this case by two points; 1. I agree that this shouldn't be a major problem while `reconnect.backoff.ms` is set to default, but still users may set `reconnect.backoff.ms` to longer value wihtout understanding how and when is it actually applied. 2. Basically we expect that "reconnect backoff" is applied for waiting before reconnecting to the server which failed to connect once. Since the `leastLoadedNode` in next round should returns different node than the one that failed to connect to, it doesn't make sense to apply reconnect backoff between connect attempts for different nodes.

What do you think?